Distributor node

Lezek123 opened the pull request:
Joystream Distributor Node (Argus)
See detailed documentation at: https://github.com/Lezek123/substrate-runtime-joystream/blob/distributor-node/distributor-node/README.md

shamil-gadelshin committed:
storage-node-v2: Initialize CLI framework (oclif).

shamil-gadelshin committed:
storage-node-v2: Add polkadot api dependencies.

shamil-gadelshin committed:
storage-node-v2: Improve ‘create storage bucket’ command.

- add prettier

shamil-gadelshin committed:
storage-node-v2: Add ‘AcceptInvitation’ command.

- Add ‘AcceptStorageBucketInvitation’ command for storage bucket providers.

shamil-gadelshin committed:
storage-node-v2: Modify extrinsic execution.

shamil-gadelshin committed:
storage-node-v2: Add linter.

shamil-gadelshin committed:
storage-node-v2: Add ‘UpdateBag’ command.

- it runs the `updateStorageBucketsForBag` extrinsic.

shamil-gadelshin committed:
storage-node-v2: Start adding ‘Upload’ command.

shamil-gadelshin committed:
storage-node-v2: Start adding ‘Dev-Init’ command.

shamil-gadelshin committed:
storage-node-v2: Add `ApiBaseCommand` type.

shamil-gadelshin committed:
storage-node-v2: Modify development mode.

shamil-gadelshin committed:
storage-node-v2: Modify ‘upload’ command.

shamil-gadelshin committed:
storage-node-v2: Refactor ‘dev-init’ command.

shamil-gadelshin committed:
storage-node-v2: Add key-file handling.

shamil-gadelshin committed:
Update storage node v2 types.

shamil-gadelshin committed:
storage-node-v2: Stat adding web API.

- add API spec
- change joystream types version
- add uploading command stub

shamil-gadelshin committed:
storage-node-v2: Lint and format.

shamil-gadelshin committed:
storage-node-v2: Run ‘acceptPendingDataObjects’ extrinsic on uploading.

shamil-gadelshin committed:
storage-node-v2: Change oclif tests.

shamil-gadelshin committed:
storage-node-v2: Add some tests.

shamil-gadelshin committed:
storage-node-v2: Add swagger editor script.

shamil-gadelshin committed:
storage-node-v2: Add command testing script.

shamil-gadelshin committed:
storage-node-v2: Refactor project structure.

shamil-gadelshin committed:
storage-node-v2: Add file hashing.

- add file hasing on upload
- add multihash support
- add blake3 support
- rename file on uploading to the multihash

shamil-gadelshin committed:
storage-node-v2: Add static files service.

shamil-gadelshin committed:
storage-node-v2: Start adding token authorization.

shamil-gadelshin committed:
storage-node-v2: Add ‘LeaderUpdateBagLimit’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘LeaderUpdateVoucherLimits’ command.

shamil-gadelshin committed:
storage-node-v2: Update runtime api usage.

- protect nonce using async lock
- use single api initialization point

shamil-gadelshin committed:
storage-node-v2: Disable prettier lints.

shamil-gadelshin committed:
storage-node-v2: Start adding bagId parser.

shamil-gadelshin committed:
storage-node-v2: Update upload method.

shamil-gadelshin committed:
storage-node-v2: Add custom logger.

- Add winston logger
- Add winston express integration

shamil-gadelshin committed:
storage-node-v2: Introduce oclif errors and exit codes.

shamil-gadelshin committed:
storage-node-v2: Change web api error handling.

shamil-gadelshin committed:
storage-node-v2: Refactor runtime api functions.

shamil-gadelshin committed:
storage-node-v2: Add custom runtime API url.

shamil-gadelshin committed:
storage-node-v2: Add parameters to the token request.

- Add storageBucketId and bagId.

shamil-gadelshin committed:
storage-node-v2: Update code comments.

shamil-gadelshin committed:
storage-node-v2: Update auth-token

Add more fields:
- storageBucketId
- bagId

shamil-gadelshin committed:
storage-node-v2: Add token request authentication.

shamil-gadelshin committed:
storage-node-v2: Add timestamp validation on upload request.

shamil-gadelshin committed:
storage-node-v2: Add token nonce and nonce local cache.

shamil-gadelshin committed:
storage-node-v2: Add static file endpoint.

shamil-gadelshin committed:
storage-node-v2: Add `delete storage bucket` command.

shamil-gadelshin committed:
storage-node-v2: Add extrinsic wrapper.

shamil-gadelshin committed:
storage-node-v2: Add ‘invite operator’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘cancel-invite’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘remove-operator’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘update data fee’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘set operator metadata’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘set uploading block’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘set storage bucket status’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘set storage bucket voucher limits’ command.

shamil-gadelshin committed:
storage-node-v2: Add ‘update dynamic bag policy’ command.

shamil-gadelshin committed:
types: Add storage DynamicBagType.

shamil-gadelshin committed:
types: Add storage DynamicBagCreationPolicy.

shamil-gadelshin committed:
types: Add storage ContentId types.

shamil-gadelshin committed:
storage-node-v2: Add ‘update blacklist’ command.

Commit:
Query node for storage_v2 - initial project setup

Commit:
Initial schema

shamil-gadelshin committed:
storage-node-v2: Add ‘storage bucket’ type.

Commit:
Merge branch 'storage_node_v2' into storagev2-query-node

shamil-gadelshin committed:
storage-node-v2: Add ‘static bag’ type.

shamil-gadelshin committed:
storage-node-v2: Add ‘dynamic bag’ types.

Commit:
Update types

Commit:
Query node: Further setup adjustments

Commit:
Query node: Initial Storage v2 mappings

Commit:
Upgrade dockerfiles / github workflows to node 14.x

shamil-gadelshin committed:
runtime: storage: Add `create distribution bucket family’ extrinsic.

Commit:
ContentId clash workaround

shamil-gadelshin committed:
runtime: Fix tests.

shamil-gadelshin committed:
runtime: storage: Add ‘MaxDistributionBucketFamilyNumber’ constant.

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- delete_distribution_bucket_family

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add ‘create_distribution_bucket’ extrinsic
- add MaxDistributionBucketNumberPerFamily constant

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `update_distribution_bucket_status` extrinsic

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `delete_distribution_bucket` extrinsic

shamil-gadelshin committed:
runtime: storage: Add exrinsic.

- add `update_distribution_buckets_for_bag`
- add `DistributionBucketsPerBagValueConstraint` constant

shamil-gadelshin committed:
runtime: storage: Fix bug.

Fix `update_storage_buckets_for_bag` extrinsic. Change bucket limit comparison.

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `distribution_buckets_per_bag_limit`

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `update_distribution_bucket_mode` extrinsic

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `update_families_in_dynamic_bag_creation_policy ` extrinsic

shamil-gadelshin committed:
runtime: storage: Add random distribution buckets pick.

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `invite_distribution_bucket_operator`

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `cancel_distribution_bucket_operator_invite`

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `accept_distribution_bucket_invitation`

shamil-gadelshin committed:
runtime: storage: Add extrinsic.

- add `set_distribution_operator_metadata`

shamil-gadelshin committed:
runtime: storage: Add constant.

- add `MaxNumberOfPendingInvitationsPerDistributionBucket`

shamil-gadelshin committed:
runtime: storage: Add some tests.

shamil-gadelshin committed:
runtime: storage: Fix comments.

Commit:
Distributor node - initial commit

Commit:
Merge remote-tracking branch 'shamil/storage-distribution' into distributor-node

Commit:
@joystream/types adjustments

Commit:
create-bucket-family, set-buckets-per-bag-limit, create-bucket, update-bag

Commit:
delete bucket family, delete-bucket, flags standarization

Commit:
update-bucket-mode, update-bucket-status

Commit:
update-dynamic-bag-policy, "multiple" flags, CONFIG_PATH env

Commit:
invite-bucket-operator, cancel-invitation

Commit:
accept-distribution-bucket-invitation

Commit:
Merge branch 'storage-distribution' into storagev2-query-node

Commit:
Types updates from `distributor-node` branch

Commit:
Add metadata-protobuf package + storage v2 metadata standards

Commit:
Add metadata-protobuf package + storage v2 metadata standards

Commit:
Distribution system initial schema and mappings

Commit:
operator:set-metadata

Commit:
Update graphql schema and queries

Commit:
Fix storage node metadata handling bug

Commit:
Request queue, improved downloads & storage node proxying, error handling, bug fixes

Commit:
CLI readme

Commit:
Cache logic, mimeType fix, requests handling optimalization

Commit:
Graceful exit, downloads concurrency limit, more logs, cache logic fixes

Commit:
API - add /status, /buckets and HEAD support for /asset

Commit:
Fix object size type bug

Commit:
init-bucket script

Commit:
Logging improvements, add winston-ES, fix latency test issues

Commit:
Merge branch 'storage_node_v2' into distributor-node

Commit:
Merge branch 'storage_node_v2' into distributor-node

Commit:
Merge branch 'storagev2-query-node' into distributor-node

Commit:
Storage node scripts - add upload.sh and modify metadata

Commit:
Merge branch 'storage_v2' into distributor-node

Commit:
New commands: remove-bucket-operator, set-bucket-family-metadata

Commit:
Query-node mappings: DistributionBucketOperatorRemoved and DistributionBucketFamilyMetadataSet event handlers

Commit:
Metadata - handling for unsetting values

Commit:
Distributed buckets - add 'all' config

Commit:
Update storage node api class

Commit:
Merge branch 'storagev2-query-node' into distributor-node

Commit:
Update with new mappings, fix distribution bucket operator status check, uncomment tmp workaround

Commit:
Fix dev:init commands typos

Commit:
@joystream/types formatting

Commit:
Merge branch 'giza-protobuf-and-query-node' into distributor-node

Commit:
Distributor node - dockerize & fix dependencies

Commit:
dsitributor-node image size optiomalization

Commit:
Allow distributor-node config through ENV + config improvements

Commit:
Fix missing openapi spec after distributor-node build

Commit:
Expose node id on /status

Commit:
ConfigParser - fix for integer/array env values, allow explicit turning off logs via env

Commit:
WorkingGroup type fix

Commit:
Support for static:wg bag ids

Commit:
HTTP keep-alive optimalization, logging improvements, setStorageNodeEndpointResponseTime bug fix, timing-allow-origin

Commit:
Merge branch 'storage_v2' into distributor-node

Commit:
connection: close on checkResponseTime

Commit:
Cache fixes

Commit:
Prettier fix

Commit:
Improve configuration & error handling

Commit:
Fix "Headers already sent" issue after res.close()

Commit:
batchUpload dev utility

Commit:
Update yarn.lock

Commit:
declare module js-image-generator

Commit:
Improve error handling (mostly query-node-related)

Commit:
Merge branch 'giza-protobuf-and-query-node' into distributor-node

Commit:
New storage node api compatibility - remove hash as identifier, use version endpoint for response time checks

Commit:
Cache cleanup, intervals configuration, duplicated objectIds in fetchSupportedDataObjects fix

Commit:
Move schemas to separate location, add docs & doc generation scripts, update autogenerated files

Commit:
Docs: commands - add index file

Commit:
Merge branch 'giza-protobuf-and-query-node' into distributor-node

Commit:
Update after config adjustments

Commit:
Support for new bucket family metadata standard

Commit:
Add md files to prettierignore

Commit:
Merge branch 'giza-protobuf-and-query-node' into distributor-node

Commit:
Add new groups to test-commands script

Commit:
Update docs

Commit:
Docs update & imporvements

Commit:
Docs: Add query-node integration note

Commit:
Close parenthesis

Commit:
Docs: commands topics

Commit:
Support different ways of providing keys

Commit:
Update autogenerated docs

Commit:
Resolve relative paths to keyfiles

Commit:
File hash check

Commit:
Merge branch 'giza-polkadot-api-update' into distributor-node

Commit:
Merge branch 'giza-protobuf-and-query-node' into distributor-node

Commit:
Update distributor node following @polkadot/api changes

Commit:
Merge branch 'giza' into distributor-node

Commit:
Fix storage node api endpoint

Commit:
Fix output bucketId in create-bucket command

Commit:
Merge branch 'giza' into distributor-node

Commit:
Small cleanup

Commit:
Merge branch 'giza' into distributor-node

Commit:
Fix distributor-node docker build, adjust @joystream/types build

shamil-gadelshin commented on distributor-node/src/services/parsers/BagIdParserService.ts:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  public parseDynamicBagId(): BagId {
+    throw new Error('Function not implemented.')
```
Should be implemented.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/command-base/default.ts:
```diff
@@ -0,0 +1,99 @@
+  async init(): Promise<void> {
+    const { configPath, yes } = this.parse(this.constructor as typeof DefaultCommandBase).flags
+    const configParser = new ConfigParserService()
+    this.appConfig = configParser.loadConfing(configPath) as ReadonlyConfig
```
A typo: loadConfig

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

Commit:
Fix resolutions

shamil-gadelshin commented on distributor-node/tsconfig.json:
```diff
@@ -0,0 +1,28 @@
+    "types" : [ "node", "mocha" ],
+    "resolveJsonModule": true,
+    "strictNullChecks": true,
+    "noUnusedLocals": false, // FIXME: Temporarly disabled during initial development
```
Is it time to restore these settings?

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/test/tsconfig.json:
```diff
@@ -0,0 +1,7 @@
+{
```
Seems like a leftover code.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/test/commands/hello.test.ts:
```diff
@@ -0,0 +1,17 @@
+import { expect, test } from '@oclif/test'
```
Seems like a leftover code.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/server/controllers/public.ts:
```diff
@@ -0,0 +1,255 @@
+
+    res.setHeader('timing-allow-origin', '*')
+
+    if (!pendingDownload && this.content.exists(objectId)) {
```
This algorithm at least partially duplicates the algorithm for `HEAD` method. Is there a way to remove this duplication? Like introducing a state enum, a state calculation function and a switch-case? Will this make the code more simple and readable?

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/server/ServerService.ts:
```diff
@@ -0,0 +1,117 @@
+import { Server } from 'http'
+
+const OPENAPI_SPEC_PATH = path.join(__dirname, '../../api-spec/openapi.yml')
+
```
`ServerService` - should be probably renamed. WebApplicationService?

Lezek123 replied:
Renamed to `HttpApiService` in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/parsers/ConfigParserService.ts:
```diff
@@ -0,0 +1,124 @@
+
+    // Override config with env variables
+    this.mergeEnvConfigWith(inputConfig)
+
```
The error message could be confusing for the config file and incorrect env variables.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96, added separate env-specific errors

shamil-gadelshin commented on distributor-node/src/services/parsers/ConfigParserService.ts:
```diff
@@ -0,0 +1,124 @@
+          .map((k) => _.camelCase(k))
+
+        const valueType = this.schemaTypeOf(configSchema, path)
+        if (valueType === undefined) {
```
Could this code be simplified? 

```
if (valueType !== undefined) {
  const value = parseValue(...) 
  _.set(config, path, value)
}
```
Probably switch-case-return could be a better option than if-else-if for the parseValue() function.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/runtime/api.ts:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  sendExtrinsic(keyPair: KeyringPair, tx: SubmittableExtrinsic<'promise'>): Promise<SubmittableResult> {
+    this.logger.info(`Sending ${tx.method.section}.${tx.method.method} extrinsic from ${keyPair.address}`)
```
This function is not aware of `Sudo` extrinsics and their messages. Will it treat them as successful in case of underlying extrinsic fail?

shamil-gadelshin replied:
Here is the example of the log message:
```
2021-10-20 08:57:36:5736 SubstrateApi info: Sending sudo.sudo extrinsic from 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY
```

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/query-node/queries/queries.graphql:
```diff
@@ -0,0 +1,78 @@
+      operatorStatus_json: { isTypeOf_eq: "StorageBucketOperatorStatusActive" }
+      operatorMetadata: { nodeEndpoint_contains: "http" }
+    }
+    limit: 9999
```
How do you get this number and what will happen on exceeding it?

Lezek123 replied:
I just assumed it is so high that it will not be exceeded. I switched to "pagination" queries in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/query-node/queries/queries.graphql:
```diff
@@ -0,0 +1,78 @@
+    }
+  }
+}
+
```
Isn't there a default limit for returning rows for these queries? I encountered 1000. What will happen after exceeding it?

Lezek123 replied:
I haven't tested setting more than 1000 providers, but it may indeed be the case. I refactored the storage providers query to use iterations of 1000 rows in https://github.com/Joystream/joystream/commit/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/distributor-node/generated/api.ts:
```diff
@@ -0,0 +1,394 @@
+/* tslint:disable */
```
Are there any references for the distributor node API?

Lezek123 replied:
Currently there are none. I added it because initially I was considering implementing distributor-to-distributor communication.

shamil-gadelshin commented on distributor-node/src/services/networking/NetworkingService.ts:
```diff
@@ -0,0 +1,350 @@
+        availabilityQueue.stop()
+        const job = async () => {
+          const api = new StorageNodeApi(endpoint, this.logging)
+          const response = await api.downloadObject(objectId, startAt)
```
What will happen on error here?

Lezek123 replied:
It will be handled by `objectDownloadQueue.on('error')`

shamil-gadelshin commented on distributor-node/src/services/networking/NetworkingService.ts:
```diff
@@ -0,0 +1,350 @@
+          const api = new StorageNodeApi(endpoint, this.logging)
+          const available = await api.isObjectAvailable(objectId)
+          if (!available) {
+            throw new Error('Not avilable')
```
Typo

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/NetworkingService.ts:
```diff
@@ -0,0 +1,350 @@
+    const details = await this.queryNodeApi.getDataObjectDetails(objectId)
+    return {
+      exists: !!details,
+      isSupported:
```
Very complex condition, it's hard to read - please, introduce some variables with descriptive names (or functions) and present the final condition as a combination of them.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/NetworkingService.ts:
```diff
@@ -0,0 +1,350 @@
+    this.stateCache = stateCache
+    this.logger = logging.createLogger('NetworkingManager')
+    this.queryNodeApi = new QueryNodeApi(config.endpoints.queryNode, this.logging)
+    // this.runtimeApi = new RuntimeApi(config.endpoints.substrateNode)
```
Please, remove commented code.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/networking/NetworkingService.ts:
```diff
@@ -0,0 +1,350 @@
+export class NetworkingService {
+  private config: ReadonlyConfig
+  private queryNodeApi: QueryNodeApi
+  // private runtimeApi: RuntimeApi
```
Please, remove commented code.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/logging/LoggingService.ts:
```diff
@@ -0,0 +1,131 @@
+        throw new Error('config.endpoints.elasticSearch must be provided when elasticSeach logging is enabled!')
+      }
+      esTransport = new ElasticsearchTransport({
+        level: config.log.elastic,
```
Perhaps introducing the `index` property would be a good idea. Currently, it will send logs to the default index. If it has a different format (e.g., for timestamps) from another application - it will fail with the `ParsingError` exception.

Lezek123 replied:
Done in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/content/ContentService.ts:
```diff
@@ -0,0 +1,269 @@
+          return
+        }
+
+        if (bytesWritten !== bytesRecieved || bytesWritten !== expectedSize) {
```
typo (received)

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/content/ContentService.ts:
```diff
@@ -0,0 +1,269 @@
+
+      pipeline(dataStream, fileStream, async (err) => {
+        const { bytesWritten } = fileStream
+        const finalHash = multihash.toB58String(multihash.encode(hash.digest(), 'blake3'))
```
Hashing algorithm is duplicated (batchUpload.ts). We could probably change hashing to the 'blake2' in the future version (because of the new Parity pallet). Perhaps having it as a dedicated helper could be a good idea.

Lezek123 replied:
Done in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96 (introduced `ContentHash` class)

shamil-gadelshin commented on distributor-node/src/services/content/ContentService.ts:
```diff
@@ -0,0 +1,269 @@
+  }
+
+  public async detectMimeType(objectId: string): Promise<string> {
+    const result = await FileType.fromFile(this.path(objectId))
```
It is a potentially very long operation. Some files have this information in the end or don't have it at all. Please, consider reading a chunk of the file. You may consult Mokhtar for details.

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96 according to storage node implementation.

shamil-gadelshin commented on distributor-node/src/services/cache/StateCacheService.ts:
```diff
@@ -0,0 +1,290 @@
+
+// LRU-SP cache parameters
+// Since size is in KB, these parameters should be enough for grouping objects of size up to 2^24 KB = 16 GB
+// TODO: Intoduce MAX_CACHED_ITEM_SIZE and skip caching for large objects entirely? (ie. 10 GB objects)
```
Is this TODO still relevant?

Lezek123 replied:
Implemented in https://github.com/Joystream/joystream/pull/2582/commits/e1d1049b589ae3f9023ac94412e72f1d82987353

shamil-gadelshin commented on distributor-node/src/services/cache/StateCacheService.ts:
```diff
@@ -0,0 +1,290 @@
+import fs from 'fs'
+
+// LRU-SP cache parameters
+// Since size is in KB, these parameters should be enough for grouping objects of size up to 2^24 KB = 16 GB
```
Do we have this as a hardware requirement for memory?

Lezek123 replied:
I'm not sure what you mean, the `16GB` only refers to the upper range of data object size in the last cache group (assuming `popularity=1`, which is the lowest possible, the 24th cache group will contain objects of size 8-16 GB, but because this is the last group, all objects of size `>16GB` will also end up in this group)

shamil-gadelshin commented on distributor-node/package.json:
```diff
@@ -0,0 +1,139 @@
+    "@elastic/ecs-winston-format": "^1.1.0",
+    "@oclif/command": "^1",
+    "@oclif/config": "^1",
+    "@oclif/plugin-help": "^2",
```
Something wrong with this dependency:
```
yarn joystream-distributor
yarn run v1.22.11
$ /root/leszek/substrate-runtime-joystream/node_modules/.bin/joystream-distributor
    Error: Could not load a help class, consider installing the @oclif/plugin-help 
    package, failed with message:
    Cannot destructure property 'underline' of 'chalk_1.default' as it is undefined.
```

Lezek123 replied:
Fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin commented on distributor-node/src/services/parsers/ConfigParserService.ts:
```diff
@@ -0,0 +1,124 @@
+import configSchema, { bytesizeUnits } from '../../schemas/configSchema'
+import { JSONSchema4 } from 'json-schema'
+
+const MIN_CACHE_SIZE = 20 * Math.pow(1024, 3)
```
How did you come to this value?

Lezek123 replied:
I used `2*MaxDataObjectSize`, which seemed to me like a reasonable minimum. Should I change it?

shamil-gadelshin commented on distributor-node/src/app/index.ts:
```diff
@@ -0,0 +1,151 @@
+        pendingDownloadsCount,
+      })
+      await new Promise((resolve) => setTimeout(resolve, 5000))
+      this.stateCache.saveSync()
```
No error handling for the IO operation. Is it intentional?

Lezek123 replied:
Any error here will be caught in https://github.com/Lezek123/substrate-runtime-joystream/blob/distributor-node/distributor-node/src/app/index.ts#L138

shamil-gadelshin commented on distributor-node/src/services/content/ContentService.ts:
```diff
@@ -0,0 +1,269 @@
+          return
+        }
+
+        const mimeType = await this.detectMimeType(objectId)
```
`detectMimeType` could throw an error that will kill the App
Here is the console log message:
```
2021-10-21 14:37:28:3728 App error: Node initialization failed!
{
    "err": {}
}
```

Here is file log postmortem:
```
{"@timestamp":"2021-10-21T14:53:26.023Z","log.level":"error","message":"Node initialization failed!","ecs":{"version":"1.6.0"},"error":{"type":"TypeError","message":"buf.readUInt8 is not a function","stack_trace":"TypeError: buf.readUInt8 is not a function\n    at Object.get (/root/leszek/substrate-runtime-joystream/distributor-node/node_modules/token-types/lib/index.js:12:20)\n    at FileTokenizer.peekNumber (/root/leszek/substrate-runtime-joystream/node_modules/strtok3/lib/AbstractTokenizer.js:63:22)\n    at readField (/root/leszek/substrate-runtime-joystream/distributor-node/node_modules/file-type/core.js:680:16)\n    at readElement (/root/leszek/substrate-runtime-joystream/distributor-node/node_modules/file-type/core.js:695:15)\n    at _fromTokenizer (/root/leszek/substrate-runtime-joystream/distributor-node/node_modules/file-type/core.js:717:14)\n    at Object.fromFile (/root/leszek/substrate-runtime-joystream/distributor-node/node_modules/file-type/index.js:8:10)\n    at ContentService.detectMimeType (/root/leszek/substrate-runtime-joystream/distributor-node/src/services/content/ContentService.ts:162:20)\n    at ContentService.startupInit (/root/leszek/substrate-runtime-joystream/distributor-node/src/services/content/ContentService.ts:104:56)\n    at App.start (/root/leszek/substrate-runtime-joystream/distributor-node/src/app/index.ts:80:7)\n    at StartNode.run (/root/leszek/substrate-runtime-joystream/distributor-node/src/commands/start.ts:13:5)"},"label":"App"}
```

Lezek123 replied:
Should be fixed in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

shamil-gadelshin reviewed (changes requested):
Awesome work! I especially liked the service structure! 

I tested single file serving in local and cross datacenter setup (distributor-node, storage-node, and browser in different locations). I verified proper response headers for 'hit', 'miss', and 'pending'. All works as expected with the exception of 'pending' described below.

I found several issues:
- If storage node stops - internal server error 500 (UnhandledPromiseRejectionWarning).
- It fails to start with an incorrect log folder. Is it intended behavior? 
- The node cannot be killed even with numerous kill signals unless it finishes its shutdown. (a stubborn one :)
- This PR is out-of-sync with the storage node metadata format but it shouldn't be a problem after the Colossus v2 merge.
- Please, add volta config to the node for consistency.
- Are there any references for the Distributor node generated API?
- `yarn checks`  produces multiple warnings

Most issues are easy to fix. However, I found one troubling issue that I'm not sure how to reproduce:
I encountered a situation that the data object is shown as `pending download` forever without changing its status. I'm not sure about the reason for it (logs are empty).
To mitigate this unknown problem I suggest introducing some timeout for files probably a long one (hours) for downloading objects.

Commit:
Address https://github.com/Joystream/joystream/pull/2582 review comments

Commit:
Implement pendingDownloadTimeout and maxCachedItemSize

Commit:
Improve error handling, logging, pending downloads flow and startup logic (drop files with missing cache data)

Commit:
Add prematurelyClosed metadata to request logs

Commit:
Fix PendingDownload cleanup on dropById

Lezek123 commented:
Thank you very much for the thorough review!

Addressing some remaining comments:

> If storage node stops - internal server error 500 (UnhandledPromiseRejectionWarning).

I improved the error handling in https://github.com/Joystream/joystream/pull/2582/commits/4171f5887f6fc589eedc09f8a77c9d69c5b10d41


> It fails to start with an incorrect log folder. Is it intended behavior?

This depends on the `logs.file` setting. If it's not missing / set to `off`, then invalid logs directory should indeed cause an error. If it's set to `off` or missing, then `directories.logs` should not matter.

> The node cannot be killed even with numerous kill signals unless it finishes its shutdown. (a stubborn one :)

Unless killed with `kill -9` (`SIGKILL`) the node will attempt to finish pending downloads (currently this should take no more than 15-20 seconds)

> Please, add volta config to the node for consistency.

Done in https://github.com/Joystream/joystream/pull/2582/commits/d42f98bc029808c0d62c02dc6880a4e338a1aa96

> Are there any references for the Distributor node generated API?

Currently there are none. I added it because initially I was considering implementing distributor-to-distributor communication.

> Most issues are easy to fix. However, I found one troubling issue that I'm not sure how to reproduce:
I encountered a situation that the data object is shown as pending download forever without changing its status. I'm not sure about the reason for it (logs are empty).

I'm not 100% sure what was the cause, but I did some refactorization in https://github.com/Joystream/joystream/pull/2582/commits/4171f5887f6fc589eedc09f8a77c9d69c5b10d41, which hopefully will solve this issue.

> To mitigate this unknown problem I suggest introducing some timeout for files probably a long one (hours) for downloading objects.

Done in https://github.com/Joystream/joystream/pull/2582/commits/e1d1049b589ae3f9023ac94412e72f1d82987353

Commit:
contentSizeSum calculation bug fix

Commit:
Update config docs

shamil-gadelshin reviewed (changes requested):
Some remaining minor issues:
- please, remove unused distribution API with documentation
- remove TODO in public.ts `// TODO: Limit the number of times useContent ...`
- remove TODO in caching or change the space number (StateCacheService.ts): `2 // TODO: Only for debugging`
- `this.stateCache.saveSync()` is not always safe
 
Here is the bug with moderate severity: distributor node fails when connected to the old storage node version (asset ID is hash-based):
```
2021-11-04 20:21:43:2143 ContentService error: Content rejected: Bytes written/received/expected mismatch!
{
    "objectId": "0",
    "expectedSize": 863,
    "bytesReceived": 853,
    "bytesWritten": 853,
    "expectedHash": "gWE26heT41YLqTiZ8QNPNxMiVpjmGxnEsSFPDewKWeQRyv",
    "finalHash": "gWC4xcyGCju8HiJwGx9bHCDsFzDZvsjDY4WPMc2k9PVnBD"
}
2021-11-04 20:21:49:2149 StateCacheService verbose: Cache file updated
2021-11-04 20:21:49:2149 NetworkingManager verbose: Checking nearby storage nodes...
{
    "validEndpointsCount": 1
}
2021-11-04 20:21:49:2149 ContentService verbose: Performing cache cleanup...
{
    "supportedObjects": 10,
    "objectsInCache": 0
}
2021-11-04 20:21:49:2149 ContentService verbose: Cache cleanup finished
{
    "droppedObjects": 0
}
2021-11-04 20:21:49:2149 NetworkingManager verbose: Mean response times updated
{
    "responseTimes": []
}
2021-11-04 20:22:08:228 ExpressServer http: HTTP GET /api/v1/asset/0
{
    "meta": {
        "req": {
            "url": "/api/v1/asset/0",
            "headers": {
                "host": "localhost:3334",
                "connection": "keep-alive",
                "sec-ch-ua": "\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"",
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": "\"macOS\"",
                "dnt": "1",
                "upgrade-insecure-requests": "1",
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.54 Safari/537.36",
                "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
                "sec-fetch-site": "none",
                "sec-fetch-mode": "navigate",
                "sec-fetch-user": "?1",
                "sec-fetch-dest": "document",
                "accept-encoding": "gzip, deflate, br",
                "accept-language": "en-GB,en-US;q=0.9,en;q=0.8,ru;q=0.7",
                "cookie": "csrftoken=lllF1wICvJ5notxdTW8vHWZoBytBrv0lL4PCGTYTV9OItDypKSWo6ShCpJ2dsQYy",
                "sec-gpc": "1"
            },
            "method": "GET",
            "httpVersion": "1.1",
            "originalUrl": "/api/v1/asset/0",
            "query": {}
        },
        "res": {
            "statusCode": 200
        },
        "responseTime": 24958,
        "prematurelyClosed": false
    }
}
2021-11-04 20:22:08:228 PublicApiController verbose: Data object requested
{
    "objectId": "gWE26heT41YLqTiZ8QNPNxMiVpjmGxnEsSFPDewKWeQRyv",
    "objectStatus": {
        "type": "NotFound"
    }
}
2021-11-04 20:22:08:228 ExpressServer http: HTTP GET /api/v1/asset/gWE26heT41YLqTiZ8QNPNxMiVpjmGxnEsSFPDewKWeQRyv
{
    "meta": {
        "req": {
            "url": "/api/v1/asset/gWE26heT41YLqTiZ8QNPNxMiVpjmGxnEsSFPDewKWeQRyv",
            "headers": {
                "host": "localhost:3334",
                "connection": "keep-alive",
                "sec-ch-ua": "\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"",
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": "\"macOS\"",
                "dnt": "1",
                "upgrade-insecure-requests": "1",
                "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.54 Safari/537.36",
                "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
                "sec-fetch-site": "none",
                "sec-fetch-mode": "navigate",
                "sec-fetch-user": "?1",
                "sec-fetch-dest": "document",
                "accept-encoding": "gzip, deflate, br",
                "accept-language": "en-GB,en-US;q=0.9,en;q=0.8,ru;q=0.7",
                "cookie": "csrftoken=lllF1wICvJ5notxdTW8vHWZoBytBrv0lL4PCGTYTV9OItDypKSWo6ShCpJ2dsQYy",
                "sec-gpc": "1"
            },
            "method": "GET",
            "httpVersion": "1.1",
            "originalUrl": "/api/v1/asset/gWE26heT41YLqTiZ8QNPNxMiVpjmGxnEsSFPDewKWeQRyv",
            "query": {}
        },
        "res": {
            "statusCode": 404
        },
        "responseTime": 59,
        "prematurelyClosed": false
    }
}
2021-11-04 20:22:11:2211 PublicApiController verbose: Data object requested
{
    "objectId": "2",
    "objectStatus": {
        "type": "Missing",
        "objectData": {
            "objectId": "2",
            "accessPoints": {
                "storageNodes": [
                    {
                        "bucketId": "0",
                        "endpoint": "http://localhost:3333/api/v1"
                    }
                ]
            },
            "contentHash": "gWCZCvZKBY1psbUi7ccK8ch5dSbYQdWsgxN1H58opwG7jT",
            "size": 881
        }
    }
}
2021-11-04 20:22:11:2211 NetworkingManager info: Downloading new data object
{
    "objectId": "2",
    "possibleSources": [
        {
            "endpoint": "http://localhost:3333/api/v1",
            "meanResponseTime": 99999
        }
    ]
}
2021-11-04 20:22:11:2211 StorageNodeApi verbose: Sending download request
{
    "0": {
        "endpoint": "http://localhost:3333/api/v1"
    },
    "objectId": "2"
}
2021-11-04 20:22:11:2211 NetworkingManager info: Download source chosen
{
    "objectId": "2",
    "source": "http://localhost:3333/api/v1"
}
2021-11-04 20:22:11:2211 ContentService verbose: Handling new content
{
    "objectId": "2",
    "expectedSize": 881
}
2021-11-04 20:22:11:2211 ContentService verbose: Reserved space for new data object
{
    "objectId": "2",
    "expectedSize": 881,
    "newContentSizeSum": 881
}
Error: Unexpected content size: Too much data received from source!
    at IncomingMessage.onData (/Users/shamix/rust/joystream/substrate-runtime-joystream/distributor-node/src/services/content/ContentService.ts:248:30)
    at IncomingMessage.emit (events.js:327:22)
    at IncomingMessage.Readable.read (internal/streams/readable.js:519:10)
    at flow (internal/streams/readable.js:992:34)
    at resume_ (internal/streams/readable.js:973:3)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)
2021-11-04 20:22:11:2211 App info: Exiting...
2021-11-04 20:22:11:2211 ExpressServer info: Express server stopped
2021-11-04 20:22:11:2211 App info: Critical exit finished
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

shamil-gadelshin commented:
Also, there is an arguable issue with REST API naming. I'd suggest renaming *asset to *assets to comply with collection names recommendation.

shamil-gadelshin commented:
Consider setting the maximum log files size (via maxSize and maxFiles parameters for logging).

Commit:
Remove unused distributor-node api

Commit:
Remove outdated TODOs

Commit:
Wrap saveSync in exitHandler in try-catch

Commit:
Update distributor node api: /asset -> /assets

Commit:
More extensive logging configuration

Commit:
Small config schema improvements, update docs

Commit:
Introduce authorized OperatorApi

Commit:
Simplify the config

Commit:
Update docs

Commit:
Fix workerId check

Commit:
formatting

Commit:
Merge branch 'giza-mappings-update-hydra' into distributor-node

Commit:
Adjustments related to https://github.com/Joystream/joystream/pull/2847

Commit:
Merge branch 'giza' into distributor-node

Commit:
Prettier - include autogenerated files

Commit:
Merge branch 'giza' into distributor-node

Commit:
Merge branch 'giza' into distributor-node

Commit:
Update distribution bucket id representation

Commit:
Fix test-commands and docker-compose example

Commit:
Improve pending download logic and error handling for "Too much data received from source" case

Commit:
Update docs

Lezek123 commented:
> Here is the bug with moderate severity: distributor node fails when connected to the old storage node version (asset ID is hash-based)

I wasn't able to reproduce this bug for a long time, but finally solved this in https://github.com/Joystream/joystream/pull/2582/commits/1f4704861beaf7d652c13b23c6d52bca008557ab.

It seems like after some initial fixes it was only occuring when expected data object size was very small (ie. few hundred bytes).

Commit:
Argus: Remove storage node auth in batchUpload command

Commit:
Merge branch 'giza' into distributor-node

Commit:
Merge branch 'giza' into distributor-node

shamil-gadelshin commented on distributor-node/src/schemas/configSchema.ts:
```diff
@@ -52,32 +48,92 @@ export const configSchema: JSONSchema4 = {
+                },
+                frequency: {
+                  description: 'The frequency of creating new log files (regardless of maxSize)',
+                  default: 'daily',
```
I think this default setting is unsafe. I recommend 'monthly' with at least 2 files or removing this config parameter.

shamil-gadelshin replied:
On the other hand - `daily` seems like the most widely used choice. I was concerned about having a small value of the files number parameter (1 or 2 days of logs) and an error week old. However, we could set a recommendation for having at least seven days of logs (with the default value >= 7).

Lezek123 replied:
Addressed in https://github.com/Joystream/joystream/pull/2582/commits/76abd622fc3dfb822c8813d8a58c6e5585a12960

shamil-gadelshin commented on distributor-node/config/docker/config.docker.yml:
```diff
@@ -0,0 +1,25 @@
+  saveCacheState: 60
+  checkStorageNodeResponseTimes: 60
+  cacheCleanup: 60
+port: 3334
```
Do we need port 3335 here as well?

Lezek123 replied:
This is already configured on the `giza_staging` branch, I'd like to avoid changing the config here, because the docker setup on `giza` branch is outdated compared to `giza_staging`, so this would cause unnecessary conflicts.

shamil-gadelshin commented on distributor-node/src/commands/node/start-public-api.ts:
```diff
@@ -0,0 +1,17 @@
+    return '/api/v1/start-api'
+  }
+
+  protected reqBody(): Record<string, unknown> {
```
Why do you need to reintroduce the same method as in the base class? Isn't the default void implementation enough?

Lezek123 replied:
Addressed in https://github.com/Joystream/joystream/pull/2582/commits/943c053358edf1766613815a7b001070cb336e9b

shamil-gadelshin commented on distributor-node/src/services/logging/LoggingService.ts:
```diff
@@ -0,0 +1,151 @@
+        level: config.logs.elastic.level,
+        format: winston.format.combine(pauseFormat({ id: 'es' }), escFormat()),
+        flushInterval: 5000,
+        source: config.id,
```
Please, add a retry limit parameter. It looks like a memory leak without it. During the invalid initialization of the elastic search server - it will constantly try to send messages many times a second.

Lezek123 replied:
Addressed in https://github.com/Joystream/joystream/pull/2582/commits/e5abadd567a52f5036c209a3ba5e2d2e0d390a58

shamil-gadelshin reviewed (changes requested):
Looks OK (and works). There are several issues left:

- operator API port 3335 doesn't seem to be reflected in the docker and docker-compose config files
- elastic search log config probably contains a memory leak without the `retry limit` parameter.

shamil-gadelshin commented:
Please, update the volta config to the latest yarn version.

Commit:
Merge branch 'giza' into distributor-node

Commit:
File logging config - change defaults and add recommendations

Commit:
Distributor node: Update yarn version in volta config

Commit:
Distributor node: Remove redundant reqBody in NodeCommandBase children

Commit:
Add elasticsearch logger retryLimit

Commit:
Update all autogenerated files

Lezek123 commented:
> Please, update the volta config to the latest yarn version.

Addressed in https://github.com/Joystream/joystream/pull/2582/commits/b6876f5ba91cb038eb01c64d4983f11829e4a5f7

Commit:
Merge branch 'giza' into distributor-node

Commit:
Fix reference to "storage-node-v2" directory

shamil-gadelshin reviewed (approved):
LGTM. Nice work!
